File: GraphClone.py

# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = []):
        self.val = val
        self.neighbors = neighbors

# # Correct, DFS Version
# class Solution:
#     def __init__(self):
#         self.map = dict()
#     def cloneGraph(self, node: 'Node') -> 'Node':
#         if not node: # deal with empty graph
#             return node
#         node_clone = Node(val=node.val)
#         # self.visited.append(node.val)
#         self.map[node] = node_clone
#         for neighbor in node.neighbors:
#             # if not (neighbor.val in self.visited):
#             if not (neighbor in self.map.keys()):
#                 neighbor_clone = self.cloneGraph(neighbor)
#                 node_clone.neighbors.append(neighbor_clone)
#             else:
#                 node_clone.neighbors.append(self.map[neighbor])
#                 # node_clone.neighbors.append(neighbor)
#         return node_clone

# Correct, BFS Version
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node: # deal with empty graph
            return node
        node_clone = Node(val=node.val)
        # map = dict()
        map = {node: node_clone}
        queue = [node]
        # queue.append(node)
        # parent = dict()
        while len(queue) > 0:
            # queue_size = len(queue)
            # for i in range(queue_size):
            current_node = queue.pop(0)
            # clone_node = Node(val=current_node.val)
            # map[current_node] = clone_node
            for neighbor in current_node.neighbors:
                if not (neighbor in map.keys()):
                    neighbor_clone = Node(val=neighbor.val)
                    map[neighbor] = neighbor_clone
                    queue.append(neighbor)
                    map[current_node].neighbors.append(neighbor_clone)
                    # parent[neighbor] = clone_node
                else:
                    map[current_node].neighbors.append(map[neighbor])
        # for child_ori in list(parent.keys()):
        #     parent_clone = parent[child_ori]
        #     parent_clone.neighbors.append(map[child_ori])

        return node_clone


    ## Wrong
# class Solution:
#     def __init__(self):
#         self.visited = []
#     def cloneGraph(self, node: 'Node') -> 'Node':
#         node_clone = Node(val=node.val)
#         self.visited.append(node.val)
#         for neighbor in node.neighbors:
#             if not (neighbor.val in self.visited):
#                 neighbor_clone = self.cloneGraph(neighbor)
#                 node_clone.neighbors.append(neighbor_clone)
#             else:
#                 node_clone.neighbors.append(neighbor)
#         return node_clone

File: GraphClone.cpp
//
// Created by Golfer on 2020/7/24.
//

#include <vector>
#include <queue>
#include <map>
#include <set>
using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;

    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }

    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }

    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

// DFS Version
class Solution {
public:
    map<Node*, Node*> visited;
    Node* cloneGraph(Node* node) {
        Node node_clone = Node(node->val);
        Node* node_clone_pointer = &node_clone;
        visited[node] = node_clone_pointer;
        for (Node* & neighbor : node->neighbors) {
//            node_clone_pointer->neighbors.resize(node_clone_pointer->neighbors.size() + 1);
            if (visited.find(neighbor) == visited.end()) {
                Node* neighbor_clone = cloneGraph(neighbor);
//                node_clone_pointer->neighbors[node_clone_pointer->neighbors.size() - 1] = neighbor_clone;
                node_clone_pointer->neighbors.push_back(neighbor_clone);
            } else{
//                node_clone_pointer->neighbors[node_clone_pointer->neighbors.size() - 1] = visited[neighbor];
                node_clone_pointer->neighbors.push_back(visited[neighbor]);
            }
        }
        return node_clone_pointer;
    }
};

int main() {
    vector<vector<int>> graph = {{2, 4}, {1, 3}, {2, 4}, {1, 3}};
    map<int, Node*> node_map;
    for (int i = 1; i <= graph.size(); i++) {
        if (node_map.find(i) == node_map.end()) {
            Node node = Node(i);
            Node* node_p = &node;
            node_map[i] = node_p;
        }
        for (int j = 0; j < graph[i-1].size(); j++) {
            if (node_map.find(graph[i][j]) == node_map.end()) {
                Node node_neighbor = Node(graph[i][j]);
                Node* node_p_neighbor = &node_neighbor;
                node_map[graph[i][j]] = node_p_neighbor;
                node_map[i]->neighbors.push_back(node_map[graph[i][j]]);
            }
            else {
                node_map[i]->neighbors.push_back(node_map[graph[i][j]]);
            }
        }
    }

    Solution solution = Solution();
    Node* clone_graph = solution.cloneGraph(node_map[0]);

    return 0;
}
